### PR TITLE
Creators and collection batch mint verifications

### PR DESCRIPTION
**Pull request by n00m4d**
# What
This PR adds creator and collection verifications to the batch mint operation.

# Why
Some of the assets in batch minting may have verified creators and collections.

Typically, creators should verify themselves by sending a transaction, which involves signature verification. In the batch minting implementation, a similar functionality is included. Here, creators must sign a hash and include this hash in the batch mint JSON file. On the indexer side, we verify the validity of the signature. If everything is correct, the creator is verified.

For collection verification, a special instruction `finalize_tree_with_root_and_collection` must be sent. The success of this transaction proves that the collection is verified by its creator. On the indexer side, we need to check if the collection in the batch mint object matches the one in the instruction.

# How
Several new checks have been added to the batch mint persister to verify creators and collections. Additionally, parsing of the `finalize_tree_with_root_and_collection` Bubblegum instruction has been implemented. The new functionality is also covered with tests.

### CODE REVIEW

File: blockbuster/blockbuster/Cargo.toml [hunk 1]
```diff
@@ -11,17 +11,17 @@ readme = "../README.md"
 [dependencies]
 bytemuck = { version = "1.14.0", features = ["derive"] }
 spl-token-2022 = { version = "1.0", features = ["no-entrypoint"] }
-spl-account-compression = { path = "../mpl-bubblegum/solana-program-library/account-compression/programs/account-compression", features = ["no-entrypoint"] }
-spl-noop = { path = "../mpl-bubblegum/solana-program-library/account-compression/programs/noop", features = ["no-entrypoint"] }
-mpl-bubblegum = { path = "../mpl-bubblegum/clients/rust" }
+spl-account-compression = { git = "https://github.com/StanChe/solana-program-library.git", branch = "feature/init_with_root", features = ["no-entrypoint"] }
+spl-noop = { git = "https://github.com/StanChe/solana-program-library.git", branch = "feature/init_with_root", features = ["no-entrypoint"] }
```
Comment (StanChe) [inline blockbuster/blockbuster/Cargo.toml:15]: you could also use workspace import here

File: blockbuster/blockbuster/src/programs/bubblegum/mod.rs [hunk 2]
```diff
@@ -348,7 +357,35 @@ fn build_create_tree_with_root_payload(
...
+    let args = FinalizeTreeWithRootAndCollectionInstructionArgs::try_from_slice(ix_data)?;
+
+    let tree_id = *keys
+        .get(1)
+        .ok_or(BlockbusterError::InstructionParsingError)?;
+    let staker = *keys
+        .get(4)
+        .ok_or(BlockbusterError::InstructionParsingError)?;
+    let collection_mint = *keys
+        .get(10)
```
Comment (StanChe) [inline blockbuster/blockbuster/src/programs/bubblegum/mod.rs:380]:
```suggestion
        .get(11)
```

**Review by StanChe — APPROVED** [review]

File: blockbuster/blockbuster/src/programs/bubblegum/mod.rs [hunk 3]
```diff
@@ -35,19 +36,22 @@ pub struct FinalizeTreeWithRootInstructionArgsWithStaker {
...
+    pub collection_mint: Option<Pubkey>,
 }
 impl FinalizeTreeWithRootInstructionArgsWithStaker {
     fn build_finalize_tree_with_root_instruction_args_with_staker(
         args: FinalizeTreeWithRootInstructionArgs,
         staker: Pubkey,
+        collection_mint: Option<Pubkey>,
     ) -> Self {
         Self {
-            rightmost_root: args.rightmost_root,
```
Comment (RequescoS) [inline blockbuster/blockbuster/src/programs/bubblegum/mod.rs:45]: lets rename to just 'root'?

File: program_transformers/src/batch_minting/tests.rs [hunk 4]
```diff
@@ -11,7 +11,7 @@ use spl_concurrent_merkle_tree::concurrent_merkle_tree::ConcurrentMerkleTree;
 use std::collections::HashMap;
 use std::str::FromStr;
 
-pub fn generate_batch_mint(size: usize) -> BatchMint {
+pub fn generate_batch_mint(size: usize, creators_verified: bool) -> BatchMint {
```
Comment (RequescoS) [inline program_transformers/src/batch_minting/tests.rs:14]: this fn also can be imported from sdk

**Review by RequescoS — APPROVED** [review]

File: program_transformers/src/batch_minting/batch_mint_persister.rs [hunk 5]
```diff
@@ -545,9 +555,39 @@ impl<T: ConnectionTrait + TransactionTrait, D: BatchMintDownloader> BatchMintPer
     }
 }
 
-pub async fn validate_batch_mint(batch_mint: &BatchMint) -> Result<(), BatchMintValidationError> {
+pub async fn validate_batch_mint(
```
[the remark this answers is not in the document]
Comment (n00m4d) [reply program_transformers/src/batch_minting/batch_mint_persister.rs:558]: yep, we have [separate ticket](https://linear.app/mplx/issue/MTG-418/use-entities-and-verifications-from-sdk-in-ref-implementation) for these changes

**Pull request merged**